enhancements/rhcos/extensions.md

```diff
@@ -0,0 +1,96 @@
+---
+title: support-for-coreos-extensions
+authors:
+  - "@cgwalters"
+reviewers:
+  - "@ashcrow"
+approvers:
+  - "@ashcrow"
+  - "@crawford"
+  - "@imcleod"
+  - "@runcom"
+creation-date: 2020-04-21
+last-updated: 2020-04-21
+status: provisional
+see-also:
+replaces:
+superseded-by:
+---
+
+# Support For CoreOS extensions
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+This enhancement proposes a MachineConfig fragment like:
+```
+extensions:
+  - usbguard
+```
+
+This is the OpenShift version of the [Fedora CoreOS extension system tracker](https://github.com/coreos/fedora-coreos-tracker/issues/401).
+
+That will add additional software onto the host - but this software will still be versioned with the host (included as part of the OpenShift release payload) and upgraded with the cluster.
+
+## Motivation
+
+We have requests to ship things like `usbguard` that are needed to meet compliance in some scenarios (and `usbguard` is very useful on bare metal), but `usbguard` also makes no sense to ship in an OS for the public cloud.
+
+The `rpm-ostree` project has had as its goal from the start to re-cast RPMs as "operating system extensions" - much like how e.g. Firefox and its extensions work.  By default it operates as a pure image system, but packages can be layered on (and overridden) client side.
+
+OpenShift is already making use of this today for the [realtime kernel](https://github.com/openshift/enhancements/blob/master/enhancements/support-for-realtime-kernel.md).
+
+We propose continuing the trail blazed with the RT kernel by adding additional RPMs to `machine-os-content` that aren't part of the base OS "image", but can be installed later.  This is how `kernel-rt` works today; we added a `kernel-rt.rpm` that is committed to the container.  In the future though, we may instead ship a separate `machine-os-extensions` container, or something else.  The only API stable piece here is the `MachineConfig` field (same as for the `kernelType: realtime`).
+
+An important way to think about this is that we can still ascribe a *single version number* to this set of content - all content is still versioned with and tested with the OS (and hence the main OpenShift 4 release image).
+
+### Goals
+
+- Add support for cluster operators (notably the MCO) to deploy non-containerized software that is maintained alongside OpenShift
+- Avoid forcing the authors of software like `usbguard` to maintain two ways to ship their software (as RPM and as a container)
+
+### Non-Goals
+
+- Direct support for installing RPMs *not* from the release image
```

Comment:
Isn't this what customers actually want?

Reply:
Some of them probably, for some things...

But trying to support it would be a huge increase in scope; for things *we* ship we know we've tested it in combination with the particular OS version.  Doing that for random 3rd party RPMs (e.g. PAM hooks or hardware raid tooling) is a totally different thing.

The other big problem domain here is: who chooses when updates for those 3rd party extensions (RPMs) happen?  The easiest would be to only trigger when the base OS updates, but...

Reply:
Agreed. Some customers will want to install RPMs, but that doesn't make it supportable, opinionated, or the right move. I'd argue most customers are more interested in a stable, upgradable OCP (which RHCOS is an important part of).